codec-dns/src/main/java/io/netty/handler/codec/dns/DefaultDnsRecordEncoder.java

```diff
@@ -49,29 +54,95 @@ public void encodeRecord(DnsRecord record, ByteBuf out) throws Exception {
         encodeName(record.hostname(), out);
     }
 
-    private void encodeRawRecord(DnsRawRecord record, ByteBuf out) throws Exception {
-        encodeName(record.name(), out);
+    private void encodeOptPseudoRecord(DnsOptPseudoRecord record, ByteBuf out) throws Exception {
+        encodeRecord0(record, out);
+        out.writeShort(0);
+    }
 
-        out.writeShort(record.type().intValue());
-        out.writeShort(record.dnsClass());
-        out.writeInt((int) record.timeToLive());
+    private void encodeOptEcsRecord(DnsOptEcsRecord record, ByteBuf out) throws Exception {
+        encodeRecord0(record, out);
+
+        int sourcePrefixLength = record.sourcePrefixLength();
+        int scopePrefixLength = record.scopePrefixLength();
+        int leftOverBits = sourcePrefixLength & PREFIX_MASK;
+        InetAddress address = record.address();
+
+        byte[] bytes = address.getAddress();
+        int addressBits = bytes.length * Byte.SIZE;
+        if (addressBits < sourcePrefixLength || sourcePrefixLength < 0) {
+            throw new IllegalArgumentException(sourcePrefixLength + ": "
+                    + sourcePrefixLength + " (expected: 0 >= " + addressBits + ')');
+        }
+
+        // See http://www.iana.org/assignments/address-family-numbers/address-family-numbers.xhtml
+        final short addressNumber;
+        if (address instanceof Inet4Address) {
+            addressNumber = 1;
+        } else if (address instanceof Inet6Address) {
+            addressNumber = 2;
+        } else {
+            throw new IllegalArgumentException("address " + address + " not supported");
+        }
+
+        int payloadLength = calculateEcsAddressLength(sourcePrefixLength);
+
+        int fullpayloadLength = 2 // OPTION-CODE
+                + 2 // OPTION-LENGTH
+                + 2 // FAMILY
+                + 1 // SOURCE PREFIX-LENGTH
+                + 1 // SCOPE PREFIX-LENGTH
+                + payloadLength; //  ADDRESS...
+
+        out.writeShort(fullpayloadLength);
+        out.writeShort(8); // This is the defined type for ECS.
+
+        out.writeShort(fullpayloadLength - 4); // Not include OPTION-CODE and OPTION-LENGTH
+        out.writeShort(addressNumber);
+        out.writeByte(sourcePrefixLength);
+        out.writeByte(scopePrefixLength); // Must be 0 in queries.
+
+        if (leftOverBits > 0) {
+            int bytesLength = payloadLength - 1;
+            out.writeBytes(bytes, 0, bytesLength);
+
+            // Pad the leftover of the last byte with zeros.
+            out.writeByte(padWithZeros(bytes[bytesLength], leftOverBits));
+        } else {
+            // The sourcePrefixLength align with Byte so just copy in the bytes directly.
+            out.writeBytes(bytes, 0, payloadLength);
+        }
+    }
+
+    // Package-Private for testing
+    static int calculateEcsAddressLength(int sourcePrefixLength) {
+        return sourcePrefixLength / Byte.SIZE + ((sourcePrefixLength & PREFIX_MASK) != 0 ? 1 : 0);
```

Comment:
`sourcePrefixLength & PREFIX_MASK` is already computed in the method where this is used (`leftOverBits`) ... can we re-use it?